Issue #196 - team 5

premik91 opened the pull request:
Email verification added
The user can verify his email when he visits his account page, he clicks the link 'Please verify your email ' and that redirects him to '/account/verification/' (as you wanted). There he can click a link that sends an email(it works, I used python debug tool 'python -m smtpd -n -c DebuggingServer localhost:1025') with generated activation key(length 77).

Commit:
Issue #196

Email address validation the beginning. Checking if going in the right
direction

Commit:
First fixes

Generating url and sending to user works. Visiting url confirmation and
setting email validation=True TODO

Commit:
Fixes

Commit:
Added email to send verification page

Commit:
Some fixes

Commit:
Renamed object

premik91 commented:
When I try to translate, every line is changed like so:

-#: .\settings.py:46
+#: settings.py:46

-#: .\account\fields.py:9
+#: account/fields.py:9

Is this okay?

mitar commented:
Please update with latest from main repository.

That change is probably OK.

Commit:
Update from main

Commit:
Translation fixes

Commit:
Translation fixes

premik91 commented:
I see there are a lot of words that aren't translated, but since there is another pull request for that I didn't touch them

Commit:
Removed fuzzy

premik91 commented:
This is also fully functional, I am just not sure you will like the Views.

mitar commented on piplmesh/account/templates/user/email_verification.html:
```diff
@@ -0,0 +1,20 @@
+{% block content %}
+<div id="main">
+    {% trans "Clicking the link below will send a confirmation mail to" %} {{ user.email }}</p>
+    <a href="{% url email_verification_send %}">{% trans "Click here to verify your email" %}</a>
```
No, every user action should be a POST request. Not GET. We should not do anything over GET. Make a form.

mitar commented on piplmesh/account/views.py:
```diff
@@ -5,12 +5,16 @@
 from django.contrib import auth, messages
 from django.contrib.auth import views as auth_views
 from django.core import exceptions, urlresolvers
+from django.core.mail import EmailMultiAlternatives
```
Import only modules.

mitar commented on piplmesh/account/templates/user/email_verification.html:
```diff
@@ -0,0 +1,20 @@
+    <div class="FAQ">
+        <div class="question">{% trans "Why should I verify my email address?" %}</div>
+        <div class="content">
+            {% trans "Verifying your email address improves the security of your Account.
```
Use `blocktrans`.

mitar commented on piplmesh/account/templates/user/email_verification.html:
```diff
@@ -0,0 +1,20 @@
+    {% trans "Clicking the link below will send a confirmation mail to" %} {{ user.email }}</p>
+    <a href="{% url email_verification_send %}">{% trans "Click here to verify your email" %}</a>
+    <br><br>
+    <div class="FAQ">
```
Lower-case CSS class names.

mitar replied:
Maybe better to have `ul`/`li` for this?

mitar replied:
Or `dl`/`dt`/`dd`? Semantically could be better?

premik91 replied:
Yes I was thinking of making an ul. I don't know, never liked dl. But can make a dl for a change :)

mitar replied:
It was just an idea. You decide.

mitar replied:
But use something semantic.

mitar commented on piplmesh/account/views.py:
```diff
@@ -5,12 +5,16 @@
 
 from pushserver import signals
 
+import random
```
Python standard modules go to the top.

mitar commented on piplmesh/account/views.py:
```diff
@@ -194,6 +200,46 @@ def dispatch(self, request, *args, **kwargs):
         return form_class(self.request.user, **self.get_form_kwargs())
 
+class EmailVerification(generic_views.TemplateView):
+        template_name = 'user/email_verification.html'
```
Indentation.

Commit:
Fixes

premik91 commented:
Fixed

Commit:
Indentation

mitar commented on piplmesh/account/templates/user/account.html:
```diff
@@ -16,5 +16,10 @@
     </div>
     <div id="password_change">
         <a href="{% url password_change %}">{% trans "Change your password" %}</a></p>
```
Ehm, this is without `<p>` at the beginning?

mitar commented on piplmesh/account/templates/user/account.html:
```diff
@@ -16,5 +16,10 @@
         <a href="{% url password_change %}">{% trans "Change your password" %}</a></p>
+        {% if not user.email_validated %}
+            <div>
+                <a href="{% url email_verification %}">{% trans "Please verify your email" %}</a>
```
Simply wrap in `<p>` instead of `<div>`, or unify somehow. So maybe move one `<div>` up?

mitar commented on piplmesh/account/templates/user/email_verification.html:
```diff
@@ -0,0 +1,22 @@
+{% block title %}{% trans "Email Verification" %}{% endblock %}
+
+{% block content %}
+<div id="main">
```
Why this div? `content` block already has a div.

Commit:
Added <p> tag

mitar commented on piplmesh/account/templates/user/email_verification.html:
```diff
@@ -0,0 +1,22 @@
+
+{% block content %}
+<div id="main">
+    {% trans "Clicking the button below will send a confirmation mail to" %} {{ user.email }}</p>
```
Use `blocktrans` and translation interpolation. Do not concatenate translations, what if in some language e-mail cannot be at the end?

mitar commented on piplmesh/account/templates/user/email_verification.html:
```diff
@@ -0,0 +1,22 @@
+        {% csrf_token %}
+        <button type="submit">{% trans "Click here to verify your email" %}</button>
+    </form>
+    <br><br>
```
Don't use `<br/>` for space. Use CSS.

premik91 replied:
Can I use inline css for this padding, because putting it in css file will make faq class less reusable.

mitar replied:
You can add bottom margin to form? ;-) E-mail verification form will be only one. ;-)

mitar commented on piplmesh/account/views.py:
```diff
@@ -1,10 +1,11 @@
 from django.contrib.auth import views as auth_views
-from django.core import exceptions, urlresolvers
+from django.core import exceptions, urlresolvers, mail
+from django.http import HttpResponseRedirect
```
Import only modules.

mitar commented on piplmesh/account/views.py:
```diff
@@ -194,6 +197,46 @@ def dispatch(self, request, *args, **kwargs):
+class EmailVerification(generic_views.TemplateView):
+    template_name = 'user/email_verification.html'
+
+def EmailVerificationSend(request):
```
`request`??? What is this?

mitar replied:
Aaa. This is a function? You are crazy? Do class-based views!

mitar replied:
I will not even look further.

premik91 replied:
This is how you do it through Django tutorials(unless they've updated it in the last 4 months) :). I knew you won't like them. Will try to do it with class based views.

mitar replied:
They definitely don't use camel-case!

premik91 replied:
Maybe not, but they certainly did views that way. Will move to class based

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -16,216 +16,358 @@ msgstr ""
-#: .\account\forms.py:40
+#: account/fields.py:15
+msgid "Callable did not return datetime.date or datetime.datetime object."
+msgstr "Klicana vrednost ni vrnila datetime.date ali datetime.datetime objekta"
```
Don't translate this. Don't translate semi-internal error messages because then it is really hard to Google. Hm, maybe this string should even not be marked as translated.

premik91 replied:
Well yeah, it makes no sense marking it for translation if we shouldn't and won't translate it :)

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -16,216 +16,358 @@ msgstr ""
-#: .\account\signals.py:11
+#: account/models.py:34
+msgid "username"
+msgstr "Uporabniško ime"
```
Don't change case.

premik91 replied:
I didn't. Don't know where this came from

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -16,216 +16,358 @@ msgstr ""
+#: account/forms.py:46
 msgid "Enter the same password as above, for verification."
-msgstr "V izogib napakam ponovno vnesi isto geslo kot zgoraj."
+msgstr "V izogib napakam ponovno vnesi enako geslo kot zgoraj."
```
Is password same or equal? I think same. So, _isto_?

premik91 replied:
Yes sorry, I started translating but than saw another pull request and reversed back(unsuccessfully).

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -16,216 +16,358 @@ msgstr ""
-#: .\account\templates\user\password_change.html.py:10
-#: .\frontend\templates\user\user.html.py:11
+#: account/views.py:214
+msgid "This message was sent to verify your email address."
```
We write _e-mail_.

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -16,216 +16,358 @@ msgstr ""
-#: .\account\templates\user\login.html.py:17
-#: .\frontend\templates\header.html.py:49
+#: account/templates/user/account.html:21
+msgid "Please verify your email"
```
_e-mail address_?

mitar commented on piplmesh/urls.py:
```diff
@@ -48,6 +48,9 @@
     url(r'^account/$', account_views.AccountChangeView.as_view(), name='account'),
     url(r'^account/password/change/$', account_views.PasswordChangeView.as_view(), name='password_change'),
+    url(r'^account/verification/$', account_views.EmailVerification.as_view(), name='email_verification'),
+    url(r'^account/verification/send/$', account_views.EmailVerificationSend, name='email_verification_send'),
```
Those two could be simply only one view. On GET you display a confirmation page, on POST you send it.

mitar commented on piplmesh/settings.py:
```diff
@@ -115,6 +115,7 @@
 SECRET_KEY = '02dl2nfiacp)87-1g2$=l@b(q5+qs^)qo=byzdvgx+35q)gw&^'
 
 EMAIL_HOST = 'localhost'
+EMAIL_PORT = 1025
```
Leave default please?

premik91 replied:
Just for testing. Will remove

mitar commented:
Please do also tests. Django supports testing of e-mail sending and so on. So you can make a test, get an e-mail, open URL, see if e-mail was activated. ;-)

Also add expiry of activation tokens.

mitar commented:
Please update from main repository.

mitar commented:
Just to make sure, I am waiting for you here, no?

Commit:
First fixes

Commit:
Update

premik91 commented:
Yes I sleep long. These are most fixes, except the class based views. And variable in translations (user.email) doesn't show up, so I am guessing something is wrong.

Commit:
Some fixes

Commit:
Fix

Commit:
Fix

premik91 commented:
I now far from what you want, but I hope closer?

mitar commented on piplmesh/account/models.py:
```diff
@@ -54,6 +54,9 @@ class User(auth.User):
     is_online = mongoengine.BooleanField(default=False)
 
+    email_validated = mongoengine.BooleanField(default=False)
+    email_activation_key = mongoengine.StringField(max_length=77)
```
You still don't store time when activation key was generated.

mitar commented on piplmesh/account/models.py:
```diff
@@ -54,6 +54,9 @@ class User(auth.User):
     connection_last_unsubscribe = mongoengine.DateTimeField()
     is_online = mongoengine.BooleanField(default=False)
 
+    email_validated = mongoengine.BooleanField(default=False)
```
Do we do verification or validation?

premik91 replied:
Hmmm good question. I don't even know where or why I came up with verification. I will change everywhere to validation.

mitar replied:
Just decide for one. :-) It can probably be also verification? What do "big sites" have?

premik91 replied:
Facebook doesn't have either(confirm mail), Steam has verify, twitter same as Facebook :D. (A confirmation message ). I think that sounds even better

mitar replied:
OK. Then let us do "confirmation" everywhere? OK?

premik91 replied:
ok

mitar commented on piplmesh/account/templates/user/email_verification.html:
```diff
@@ -0,0 +1,21 @@
+
+{% block content %}
+    <p>
+        {% blocktrans %}Clicking the button below will send a confirmation mail to {{ user.email }}{% endblocktrans %}
```
This is wrong use of `blocktrans`, read documentation. ;-)

mitar commented on piplmesh/account/views.py:
```diff
@@ -1,10 +1,10 @@
-import urllib
+import datetime, random, string, urllib
```
For what do you use `datetime`?

mitar commented on piplmesh/account/views.py:
```diff
@@ -150,7 +150,9 @@ def form_valid(self, form):
         user = self.request.user
         user.first_name=form.cleaned_data['first_name']
```
Uh, why are there no spaces around `=`? Please fix also this. ;-)

mitar commented on piplmesh/account/views.py:
```diff
@@ -198,6 +200,45 @@ def dispatch(self, request, *args, **kwargs):
     def get_form(self, form_class):
         return form_class(self.request.user, **self.get_form_kwargs())
 
+class emailVerification(generic_views.TemplateView):
```
Start with capital letter for class names.

premik91 replied:
I thought that was camel-case, that's why I changed it :D

mitar commented on piplmesh/account/views.py:
```diff
@@ -198,6 +200,45 @@ def dispatch(self, request, *args, **kwargs):
+        to = user.email
+
+        # Message content
+        text_content = _("This message was sent to verify your e-mail address.")+'\n\n'
```
Create plain text Django template and render this message as template. Do not hard-code.

mitar commented on piplmesh/account/views.py:
```diff
@@ -198,6 +200,45 @@ def dispatch(self, request, *args, **kwargs):
+        user.email_activation_key = activation_key
+        user.save()
+
+        msg = mail.EmailMultiAlternatives(subject, text_content, from_email, [to])
```
There is one method on `user` object to send messages to her.

mitar commented on piplmesh/account/views.py:
```diff
@@ -198,6 +200,45 @@ def dispatch(self, request, *args, **kwargs):
+    template_name = 'user/account.html'
+
+    def get(self, request, *args, **kwargs):
+        if self.kwargs['activation_key'] == request.user.email_activation_key:
```
No, you should not allow verification through `get`!

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -17,187 +17,246 @@ msgstr ""
-#: .\frontend\templates\user\user.html.py:11
+#: account/views.py:207
+msgid "Verify your e-mail address"
+msgstr "Prosimo potrdite svoj e-mail naslov"
```
_Potrdi svoj e-mail naslov_

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -17,187 +17,246 @@ msgstr ""
+
+#: account/views.py:213
+msgid "This message was sent to verify your e-mail address."
+msgstr "To sporočilo je bilo poslano, da potrdite svoj e-mail naslov."
```
_potrdiš_

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -17,187 +17,246 @@ msgstr ""
+
+#: account/views.py:214
+msgid "Please click the link below to verify your e-mail address:"
+msgstr "Prosimo izberite spodnjo povezavo za potrditev vašega e-mail naslova:"
```
_izberi_, _tvojega_

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -17,187 +17,246 @@ msgstr ""
+
+#: account/views.py:237
+msgid "You have successfully verified your e-mail address"
+msgstr "Odjava uspešna."
```
???

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -17,187 +17,246 @@ msgstr ""
+"address'"
+msgstr ""
+"Vaša potrditvena koda je napačna. Prosimo, da izberete 'Prosimo potrdite "
+"svoj e-mail naslov' povezavo"
```
_Tvoja._

mitar replied:
And similar the rest.

Commit:
Fixes

Commit:
Fix

premik91 commented:
Some fixes, don't comment yet. I just wanted you to know I am working. :)

Commit:
Some more fixes

The rest I will try to do tomorrow, but have a lot of other homework
and two exams next week(one on monday).

premik91 commented:
The rest I will try to do tomorrow, but have a lot of other homework
and two exams next week(one on monday).

Commit:
Fixes

Commit:
Fix

premik91 commented:
Most things fixed, I have problems in 'converting' from get to post and isValid() function doesn't work as I've read it should(DateTimeField and datetime isn't the same).

Commit:
Empty line

mitar commented:
Don't forget about tests!

mitar commented on piplmesh/account/models.py:
```diff
@@ -54,10 +54,17 @@ class User(auth.User):
     is_online = mongoengine.BooleanField(default=False)
 
+    email_confirmed = mongoengine.BooleanField(default=False)
+    email_activation_key = mongoengine.StringField(max_length=77)
```
No, MongoDB supports rich schema, you can create an embedded document type for activation key and use this here. So then you would access it as `user.email_confirmation_token.value` and `user.email_confirmation_token.created_time`. Use `token` instead of `key`, no? I think it is better.

mitar replied:
Then you can also define those validity method on that embedded document! Which is much better place!

```
user.email_confirmation_token.is_valid()
```

mitar replied:
What do we do? Activation or confirmation? ;-)

mitar replied:
I edited my comments above. Please read them on GitHub.

mitar commented on piplmesh/account/models.py:
```diff
@@ -54,10 +54,17 @@ class User(auth.User):
 
+    email_confirmed = mongoengine.BooleanField(default=False)
+    email_activation_key = mongoengine.StringField(max_length=77)
+    email_activation_key_validity = models.DateTimeField()
```
Just have a creation timestamp. You can always calculate the validity based on that, especially if validity length is a settings option, so that even if settings option changes, it will get into the effect for all tokens issued.

mitar commented on piplmesh/account/models.py:
```diff
@@ -54,10 +54,17 @@ class User(auth.User):
     def get_absolute_url(self):
         return ('profile', (), {'username': self.username})
 
+    def email_activation_key_isValid(self):
```
Please name methods correctly.

mitar commented on piplmesh/account/models.py:
```diff
@@ -54,10 +54,17 @@ class User(auth.User):
         return ('profile', (), {'username': self.username})
 
+    def email_activation_key_isValid(self):
+        return (datetime.datetime.today() - self.email_activation_key_validity).days
```
Use `now()` (to get datetime) and not just `today()` (which gives just date).

Better is even to use `timezone.now()` from Django utils.

mitar replied:
Define a setting for number of days and use this here. And return only `True` of `False`.

mitar commented on piplmesh/account/templates/confirmationemail.txt:
```diff
@@ -0,0 +1,9 @@
```
Move it to proper subdirectory. Rename to `confirmation_email.txt`.

mitar replied:
See this [template](https://github.com/wlanslovenija/nodewatcher/blob/master/nodewatcher/frontend/templates/registration/activation_email.txt).

mitar commented on piplmesh/account/templates/confirmationemail.txt:
```diff
@@ -0,0 +1,9 @@
+This message was sent to you to confirm this e-mail address.
+
+Please click the link below to confirm your e-mail address:
+https://www.current_site.com/account/confirmation/{{ activation_key }}/
```
This URL does not look correct, does it? ;-)

premik91 replied:
Yes I tried to get the base url, but there were some kind of errors, so I left it like this for now.

mitar commented on piplmesh/account/templates/confirmationemail.txt:
```diff
@@ -0,0 +1,9 @@
+{% load i18n %}
+
+{% blocktrans with username=username activation_key=activation_key %}
```
Maybe split this into two/multiple blocks (each paragraph into one).

mitar commented on piplmesh/account/views.py:
```diff
@@ -1,10 +1,12 @@
 from django.contrib.auth import views as auth_views
-from django.core import urlresolvers
+from django.core import exceptions, urlresolvers
+from django.template import Context, loader
```
Import only modules!!!

mitar commented on piplmesh/account/views.py:
```diff
@@ -198,6 +202,48 @@ def dispatch(self, request, *args, **kwargs):
+    template_name = 'user/email_confirmation.html'
+
+    def post(self, request, *args, **kwargs):
+        subject = _("Confirm your e-mail address")
```
No, also subject should be in the template. See [here](https://github.com/wlanslovenija/nodewatcher/tree/master/nodewatcher/frontend/templates/registration).

mitar commented on piplmesh/account/views.py:
```diff
@@ -198,6 +202,48 @@ def dispatch(self, request, *args, **kwargs):
+        messages.success(request, _("Email confirmation link has been sent to the e-mail address you provided."), fail_silently=True)
+        return http.HttpResponseRedirect(urlresolvers.reverse_lazy('account'))
+
+class EmailConfirmationActivate(generic_views.TemplateView):
```
Maybe don't use TemplateView but some other class-based view?

Commit:
First fixes -> not finished yet

Commit:
Fixes

mitar commented on piplmesh/account/models.py:
```diff
@@ -54,10 +54,16 @@ class User(auth.User):
     is_online = mongoengine.BooleanField(default=False)
 
+    email_confirmed = mongoengine.BooleanField(default=False)
+    email_confirmation_token = mongoengine.StringField(max_length=77)
```
Hm, still no embedded document here? ;-)

premik91 commented:
Yes this and post view are still missing. Will do, tomorrow night :)

mitar commented:
You will have to pull from the main repository new changes.

Commit:
Update from main

Commit:
Mongoengine Embedded Document for Email confirmation

premik91 commented:
Something doesn't work though, I am not saving or calling it right?
<code>email_confirmation_token - 'unicode' object has no attribute 'get'<code>

mitar commented:
You probably have to flush your database. ;-)

Then, you access value through `email_confirmation_token.value`, without get (or with `getattr`).

Commit:
Fixes

premik91 commented:
Yes I just needed to flush the database. It now works.  :)
The problem is still just the get view, I am unable to change it to post. The problem is, it doesn't load the neccessary data to display the account page(I think).

mitar commented on piplmesh/account/views.py:
```diff
@@ -327,6 +330,50 @@ def dispatch(self, request, *args, **kwargs):
+            if not user.email_confirmation_token_is_valid():
+                messages.error(request, _("The confirmation code has expired. Please click 'Please confirm your e-mail address'"), fail_silently=True)
+            else:
+                user.email_confirmed = True
```
No, you should not change objects in get!

mitar commented on piplmesh/account/views.py:
```diff
@@ -327,6 +330,50 @@ def dispatch(self, request, *args, **kwargs):
     def get_form(self, form_class):
         return form_class(self.request.user, **self.get_form_kwargs())
 
+class EmailConfirmation(generic_views.TemplateView):
```
You need to views! One to send the token. The other to go to with the token and confirm. Use `FormView`, I suggest!

premik91 replied:
There already are. If it is post it sends an email and redirects him to account page, if get redirects him to email_confirmation.html. It works as expected.

mitar replied:
OK, but still use FormView to use its workflow. ;-)

But you said that something does not work for you?

mitar commented on piplmesh/account/views.py:
```diff
@@ -327,6 +330,50 @@ def dispatch(self, request, *args, **kwargs):
+class EmailConfirmationActivate(generic_views.TemplateView):
+    template_name = 'user/account.html'
+
+    def get(self, request, *args, **kwargs):
```
In get you should just prefill the form from GET and invite user to press a button to make a POST on the form.

Commit:
Fixed post and get views

premik91 commented:
Ok, both views now have get and post. It also all works(but that is usually not the problem) as you wanted. First click redirects to confirm page another sends email and redirects back to account and displays message that sending was successful. Clicking on link in mail opens a new page confirmation_final and clicking a button there redirects to account and displays message.

Commit:
Updated translations

premik91 commented:
No now it works :). I had problems, because I tried to make the confirmation link with only post(without confirmation_final page, and that of course didn't work, because the neccesary data wasn't loaded).

mitar commented:
OK. Merge with current code from main repository. And really, try to do it through FormView. ;-) I am not sure if it will be simpler, but it will be cleaner.

premik91 commented:
Trying trying. Definitely not simpler :)

mitar commented on piplmesh/account/templates/user/email_confirmaton_final.html:
```diff
@@ -0,0 +1,15 @@
+<p>
+    {% blocktrans %}You are just one step away from confirming your e-mail. Just click the button bellow.{% endblocktrans %}
+</p>
+<form method="post" action=".">
```
Hm, the idea is to open him a page with input box where token is prefllled in. Reason is that sometimes user will need to paste in the token because it does not work. So define form for this, use FormView, validate it, set flag.

Commit:
Updated from main

mitar commented:
Yes, but you will learn a lot. ;-) Check how we are doing it elsewhere.

premik91 commented:
Doing that. Ok will change to FormView :)

Commit:
Major fixes(hope the right ones)

premik91 commented:
Ready for another set of comments :)

mitar commented on piplmesh/account/forms.py:
```diff
@@ -130,3 +130,21 @@ class PasswordChangeForm(UserCurrentPasswordForm, UserPasswordForm):
+    Class with form for sending a confirmation e-mail.
+    """
+
+class EmailConfirmationActivateForm(forms.Form):
```
Confirmation and activate? ;-)

Maybe `EmailConfirmationProcessTokenForm`? And the above `EmailConfirmationSendTokenForm`?

premik91 replied:
Haha, much better yes. :)

mitar commented on piplmesh/account/forms.py:
```diff
@@ -130,3 +130,21 @@ class PasswordChangeForm(UserCurrentPasswordForm, UserPasswordForm):
+        min_length=77,
+        max_length=77,
+        widget=forms.TextInput,
+        help_text=_("Please paste in the confirmation token we sent to your e-mail."),
```
_Please enter the confirmation token you received to your e-mail address._

mitar commented on piplmesh/account/forms.py:
```diff
@@ -130,3 +130,21 @@ class PasswordChangeForm(UserCurrentPasswordForm, UserPasswordForm):
+        label=_("Confirmation token"),
+        min_length=77,
+        max_length=77,
+        widget=forms.TextInput,
```
This is default, no?

Probably you also want to add `required` flag?

mitar commented on piplmesh/account/models.py:
```diff
@@ -28,6 +28,10 @@ class Connection(mongoengine.EmbeddedDocument):
 
+class EmailConfirmationToken(mongoengine.EmbeddedDocument):
+    value = mongoengine.StringField(max_length=77)
+    date_created = mongoengine.DateTimeField()
```
```
created_time = mongoengine.DateTimeField(default=lambda: timezone.now(), required=True)
```

mitar commented on piplmesh/account/models.py:
```diff
@@ -63,10 +67,16 @@ class User(auth.User):
     def get_absolute_url(self):
         return ('profile', (), {'username': self.username})
 
+    def email_confirmation_token_is_valid(self):
```
You should move this to `EmailConfirmationToken`, no?

mitar commented on piplmesh/account/views.py:
```diff
@@ -327,6 +330,59 @@ def dispatch(self, request, *args, **kwargs):
+        user = self.request.user
+        if form.cleaned_data['confirmation_token'] == user.email_confirmation_token.value:
+            if not user.email_confirmation_token_is_valid():
+                messages.error(self.request, _("The confirmation code has expired. Please click 'Please confirm your e-mail address'"), fail_silently=True)
```
Don't reveal the reason. Just write something like "The confirmation code is invalid or has expired. Please retry."

Also, I would use this in form validation. Not here. This is the idea. Form validates that confirmation key is OK and if it is not, it is a form error. Then here you just set `email_confirmed` to true and this is it. ;-)

premik91 replied:
Agreed, will do tomorrow morning.

Commit:
Renaming

Commit:
Fixes

Commit:
Fix

premik91 commented:
Done :)

Commit:
Fix

Commit:
Final fixes

mitar commented:
Merge with latest from main repository.

mitar commented on piplmesh/account/forms.py:
```diff
@@ -130,3 +130,36 @@ class PasswordChangeForm(UserCurrentPasswordForm, UserPasswordForm):
+
+class EmailConfirmationProcessTokenForm(forms.Form):
+    """
+    Class with form for sending a confirmation e-mail.
```
I think you mixed form names around. This form should be for sending token (as comment says) and lower one for processing.

Also comment could be better: _Form for sending an e-mail address confirmation token._

premik91 replied:
omg, I did mix them.

mitar commented on piplmesh/account/forms.py:
```diff
@@ -130,3 +130,36 @@ class PasswordChangeForm(UserCurrentPasswordForm, UserPasswordForm):
+
+class EmailConfirmationSendTokenForm(forms.Form):
+    """
+    Class with form for confirming e-mail with token.
```
_Form for processing an e-mail address confirmation token._

mitar commented on piplmesh/account/forms.py:
```diff
@@ -130,3 +130,36 @@ class PasswordChangeForm(UserCurrentPasswordForm, UserPasswordForm):
+
+        confirmation_token = self.cleaned_data['confirmation_token']
+        if self.user.email_confirmation_token.value != confirmation_token or \
+           not self.user.email_confirmation_token.email_confirmation_token_is_valid():
```
`email_confirmation_token_is_valid` -> `is_valid`. ;-)

mitar commented on piplmesh/account/models.py:
```diff
@@ -28,6 +28,13 @@ class Connection(mongoengine.EmbeddedDocument):
     channel_id = mongoengine.StringField()
 
+class EmailConfirmationToken(mongoengine.EmbeddedDocument):
+    value = mongoengine.StringField(max_length=77)
```
`required=True`?

premik91 replied:
Yes will, I just thought it wasn't necessary. But it definitely doesn't hurt.

mitar commented on piplmesh/account/models.py:
```diff
@@ -28,6 +28,13 @@ class Connection(mongoengine.EmbeddedDocument):
+    value = mongoengine.StringField(max_length=77)
+    created_time = mongoengine.DateTimeField(default=lambda: timezone.now(), required=True)
+
+    def email_confirmation_token_is_valid(self):
```
I would rename this to `is_valid` or even `check_token`, which would receive given token, and do all checks (if it is same, if it has not expired). So code in the form can be simplified. And logic about what is content of the token could be here, not in the form.

Commit:
Update from main

mitar commented on piplmesh/account/models.py:
```diff
@@ -28,6 +28,13 @@ class Connection(mongoengine.EmbeddedDocument):
+    created_time = mongoengine.DateTimeField(default=lambda: timezone.now(), required=True)
+
+    def email_confirmation_token_is_valid(self):
+        return (timezone.now() - self.created_time).days < 2
```
Define a top-level constant variable for this. Give 5 days.

mitar replied:
Write about this 5 days in the confirmation e-mail (so that they know how fast they should be).

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,6 @@
+This message was sent to you to confirm this e-mail address.{% endblocktrans %}
+{% blocktrans with confirmation_token=confirmation_token %}
+Please click the link below to confirm your e-mail address:
+https://www.piplmesh.com/account/confirmation/{{ confirmation_token }}/
```
You have to fix this. ;-)

mitar replied:
Also write an alternative, something like "If link above does not work, you can open ... directly and copy-past the following code ..."

premik91 replied:
Is there any simple way of getting our base url that we already have? Or should I make settings.BASE_URL?

mitar replied:
Check how we are doing it elsewhere. For example, how we generate Facebook callback URLs. ;-)

premik91 replied:
<code>If the link above does not work, you can copy-paste it directly into your browser. </code>
This okay?

mitar replied:
No, this is one thing. To try to click, or copy-paste link. But then also copy-paste the token itself into the form.

mitar replied:
(BTW, form should be accessible also without given token.)

premik91 replied:
Yes I just added that also. Hmmm, than confirmation/{{token}} isn't really the best url. Can I change it, so /confirmation/ and this are totally different urls?

premik91 replied:
Or one can be /confirmation/ the other /confirmation/token/{{token}}
Where {{token}} isn't necessary.

mitar replied:
Why not `/confirmation/{{ token }}/` and `/confirmation/`?

premik91 replied:
/confirmation/{{ token }}/ and /confirmation/ are different views and templates. So accessing /confirmation/ would load the template to send email and /confirmation/{{ token }}/ would load token input view. 

Hmmm, putting both on one template? That would be almost complete restructure :). Should I do that?(tomorrow after the exam) :)

mitar commented on piplmesh/account/templates/user/confirmation_email_subject.txt:
```diff
@@ -0,0 +1 @@
+{% load i18n %}{% blocktrans with site_name=site_name %}[{{ site_name }}] Confirm your e-mail address{% endblocktrans %}
```
There is some default subject prefix in settings.

Commit:
First fixes

premik91 commented:
So? Are we understanding each other correctly? Because the form for token input is not really accessible without a token in url(but it could be like/confirmation/token/, but this is still a token 'token')

mitar commented:
Aha, OK, I missed that you are already using `/confirmation/` for something else. I would suggest then:
- `/confirmation/`
- `/confirmation/code/` and user has to enter the token
- `/confirmation/code/.../` to prefill it

OK, one more thing. Do we use "token" or "code"? Because in code we are talking about "token", but in user strings you are using "code". This should be unified. Decide for one.

premik91 commented:
I was using code, but than we changed views to token, and everything became confusing :). Will put token, I just thought a 'normal' user would understand code better than token.

mitar commented:
Just decide for one. ;-) I think token is quite understandable for English-spoken people. But how we will translate it to Slovene? Maybe code is better. ;-)

Commit:
Fix

Commit:
Translated

premik91 commented:
Changed to token(translation is 'koda', but it could be 'zeton' or something else, but I think 'koda' will suffice). :)

mitar commented:
Why not _code_ in English, too, then?

premik91 commented:
I don't know. I think you were right at the beginning. Token sounds much better :)

premik91 commented:
But I can put 'zeton', if you think it's better. It might be good also

premik91 commented:
I also used two mails. jon.pre... and iApremik... :) (just seen your mail)

Commit:
Update from main

mitar commented:
OK, leave _token_ and _koda_.

mitar commented on piplmesh/account/forms.py:
```diff
@@ -132,3 +132,35 @@ class PasswordChangeForm(UserCurrentPasswordForm, UserPasswordForm):
+        confirmation_token = self.cleaned_data['confirmation_token']
+        if not self.user.email_confirmation_token.check_token(confirmation_token):
+            raise forms.ValidationError(_("The confirmation token is invalid or has expired. Please retry."), code='confirmation_token_incorrect')
+        return True
```
Read what this method has to return:

https://docs.djangoproject.com/en/1.4/ref/forms/validation/

mitar commented on piplmesh/account/models.py:
```diff
@@ -19,6 +19,8 @@
 
 LOWER_DATE_LIMIT = 366 * 120
 USERNAME_REGEX = r'[\w.@+-]+'
+DAYS_TOKEN_IS_VALID = 5
```
`CONFIRMATION_TOKEN_VALIDITY = 5 # days`

mitar commented on piplmesh/account/models.py:
```diff
@@ -31,6 +33,13 @@ class Connection(mongoengine.EmbeddedDocument):
+    created_time = mongoengine.DateTimeField(default=lambda: timezone.now(), required=True)
+
+    def check_token(self, confirmation_token):
+        return ((timezone.now() - self.created_time).days < DAYS_TOKEN_IS_VALID) and confirmation_token == self.value
```
You could inverse checks and split it into multiple lines.

```
if token not equals correct value:
    return False
elif token has expired:
    return False
else:
    return True
```

It is much easier to understand conditions token has to satisfy.

mitar commented on piplmesh/account/templates/user/confirmation_email_subject.txt:
```diff
@@ -0,0 +1 @@
+{% load i18n %}{% blocktrans with site_name=site_name %}{{ site_name }} Confirm your e-mail address{% endblocktrans %}
```
No, not good. See how other subjects in Django are made. Especially see this setting:

https://docs.djangoproject.com/en/dev/ref/settings/#email-subject-prefix

mitar replied:
nodewatcher is not always the best source:

https://github.com/wlanslovenija/nodewatcher/blob/master/nodewatcher/frontend/templates/registration/activation_email_subject.txt

:-)

premik91 replied:
This is not correct?
<code>Subject: [PiplMesh] Confirm your e-mail address</code>

mitar replied:
I would do:

```
{% load i18n %}{{ EMAIL_SUBJECT_PREFIX }}{% blocktrans %}Confirm your e-mail address{% endblocktrans %}
```

Pass to the context:
- SITE_NAME (somebody could like to write that in the subject)
- EMAIL_SUBJECT_PREFIX
- user object (somebody could like to address the user in the subject)
- request object (to be able to maybe display the IP of the request)
- CONFIRMATION_TOKEN_VALIDITY (to tell how long you have to respond, in the subject)

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,10 @@
```
Some ideas for the content:

https://github.com/wlanslovenija/nodewatcher/blob/master/nodewatcher/frontend/templates/registration/activation_email.txt

mitar replied:
There is nothing wrong with an empty line here and there. ;-)

mitar replied:
Some text about what this e-mail is is missing (for those who might now have registered).

Also some footer should be made, with link to the site and so on. So that people know from who this e-mail is.

mitar replied:
So restructure this e-mail a bit. Now you added some parts and some parts are a bit confusing now.

mitar replied:
OK, you should also rewrap those translated strings. I talked about this in one prior comment.

premik91 replied:
Yes I did try that earlied, but than somehow forgot. Fixed

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,10 @@
+Token={{ confirmation_token }}
+{% endblocktrans %}
+{% blocktrans with site_url=site_url %}If link above does not work, you can copy-paste it directly into your browser.{% endblocktrans %}
+{% blocktrans %}The link will be active for 5 days, if it expires you will have to redo the process of confirming your e-mail address.{% endblocktrans %}
```
Don't hard-code this. Pass in your top-level variable.

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,10 @@
+This message was sent to you to confirm this e-mail address.{% endblocktrans %}
+{% blocktrans with confirmation_token=confirmation_token %}
+Please click the link below to confirm your e-mail address:
+{{ site_url }}{{ confirmation_token }}/
```
Don't construct URL manually like this. Read (documentation or our code) how absolute (full) URLs are made.

mitar replied:
(Again, nodewatcher is not good source here. We are using one library there which does not pass everything necessary to template context, so it is not possible to do it properly. But here we can do it properly.)

premik91 replied:
I just thought it was better to send these two variables like this, because I use both of them twice. So sending 3 variables instead of two seems wrong :)

mitar replied:
It is nothing wrong to have a rich context because then users (people who design templates) can play more and maybe do in the template something you have not envisioned. I really hate when I do not have everything I need in the template.

But yes, there is no need to put URL in the context, just put everything you need there to be able to construct the URL from the token.

So I suggest:
- SITE_NAME (we should define settings.SITE_NAME for this)
- user object
- request object
- confirmation_token
- CONFIRMATION_TOKEN_VALIDITY (capital letters all, because it is a constant)

Then you can use `url` tag (Django built-in) and `fullurl` (from [django-missing](http://django-missing.readthedocs.org/en/latest/templatetags.html#missing.templatetags.url_tags.fullurl)) to create everything else you need.

premik91 replied:
is this one needed? <code>confirmation_token</code>
 It is in user object isn't it?

mitar replied:
Good catch. But now you have to think about common users (deployers of this portal) who would like to change template a bit. It is good that they do not have to know all the details about where everything is stored. To ease them (and because the whole e-mail is exactly about this confirmation) I would leave this in. Because it is easier to use. Now that I think about, I would even add user's email address we are confirming to the template directly.

This also helps later on, if we decide to change internal implementation where we are storing tokens and so on (like we could store multiple tokens for each request ...). So it is better to give exactly the information this confirmation is about: token, e-mail address and user for which it is.

mitar commented on piplmesh/account/templates/user/email_confirmation.html:
```diff
@@ -0,0 +1,25 @@
+        <p id="current_item"><a href="{% url account %}">{% trans "Account" %}</a></p>
+    </div>
+    <p>
+        {% blocktrans with email=user.email %}Clicking the button below will send a confirmation mail to {{ email }}{% endblocktrans %}
```
Dot is missing at the end of the sentence?

mitar commented on piplmesh/account/templates/user/email_confirmation.html:
```diff
@@ -0,0 +1,25 @@
+    <dl class="faq">
+        <dt>{% trans "Why should I confirm my e-mail address?" %}</dt>
+        <dd>
+            {% blocktrans %}Confirming your e-mail address improves the security of your Account. Once you've confirmed your e-mail address, you will be able to reset your password.{% endblocktrans %}
```
_Account_ -> _account_
_reset your password_ -> *reset your password in the case you forget it."

premik91 replied:
Is it _in case_ or _in the case_?

mitar replied:
Don't know exactly, I think with _the_, but do however you want. We will give this later on to somebody to check. ;-) For now just to have the content we would like to have, this we can improve later.

mitar commented on piplmesh/account/templates/user/email_confirmation.html:
```diff
@@ -0,0 +1,25 @@
+
+{% load i18n %}
+
+{% block title %}{% trans "E-mail Confirmation" %}{% endblock %}
```
Do we capitalize every word in titles or just the first one? Here you do, bellow you don't. How are we doing elsewhere?

premik91 replied:
Only first one. :)

mitar commented on piplmesh/account/templates/user/email_confirmaton_final.html:
```diff
@@ -0,0 +1,20 @@
+        <p id="current_item"><a href="{% url account %}">{% trans "Account" %}</a></p>
+    </div>
+    <p>
+        {% blocktrans %}You are just one step away from confirming your e-mail. The confirmation token was sent to the e-mail you provided. If for some reason the input box below is empty paste the token in manually{% endblocktrans %}
```
_confirming your e-mail_ -> _confirming your e-mail address_

mitar replied:
OK, we should decide if we are confirming e-mail or confirming e-mail address. And should stick to this terminology everywhere. (In user-visible text.)

mitar replied:
_the e-mail you provided_ -> not really, there was nothing like that in the form. We could just write out again the e-mail address of the user.

mitar replied:
_If for some reason the input box below is empty, paste in the token manually._

Comma, dot at the end.

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -6,206 +6,317 @@ msgid ""
+#: account/forms.py:163
+msgid "The confirmation token is invalid or has expired. Please retry."
+msgstr ""
+"Potrditvena koda je napačna ali pa je potekla. Prosimo, da poskusiš znova."
```
_Prosimo, poskusi znova._

Commit:
Fixes

Commit:
Fixes

premik91 commented:
Done

Commit:
Added dot

mitar commented on piplmesh/account/templates/user/confirmation_email_subject.txt:
```diff
@@ -0,0 +1 @@
+{% load i18n %}{% blocktrans with site_name=site_name %}{{ site_name }}Confirm your e-mail address{% endblocktrans %}
```
OK, improve the subject how we discussed it elsewhere.

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,28 @@
+
+{% blocktrans with confirmation_token=confirmation_token %}Please click the link below to confirm your e-mail address:
+
+{{ site_url }}{{ confirmation_token }}/
```
Bljak!

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,28 @@
+{% blocktrans %}This message was sent to you to confirm this e-mail address.
+If you have already done so or don't want to, you can safely ignore this e-mail.{% endblocktrans %}
+
+{% blocktrans with confirmation_token=confirmation_token %}Please click the link below to confirm your e-mail address:
```
Don't have empty lines in blocktrans. Just add this section until the end. And then token on its own.

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,28 @@
+{% load i18n %}{% blocktrans with username=username %}Hello {{ username }}!{% endblocktrans %}
+
+{% blocktrans %}This message was sent to you to confirm this e-mail address.
```
For templates for e-mails, be careful to wrap lines correctly. Hm, is there maybe some tag to do it for you? So that output is nicely formatted, but with not too long lines.

mitar replied:
_confirm this e-mail address._ -> confirm for what? An user could forgotten what is this about. Or some other recipient could not know. Maybe better something like `confirm this e-mail address as yours for your account on {{ site_name }} with username {{ user.username }}`

mitar replied:
Aha, `filter` tag with `wrap` argument. This would do the trick.

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,28 @@
+
+{{ site_url }}{{ confirmation_token }}/
+
+Or paste the following token into the input box here:
```
No, bad instructions. Don't write anything else on the same line as token. Because people will paste everything in.

Better would be:

```
Or, open the following URL:

...

and copy-paste the following confirmation token into the input box there:

...
```

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,28 @@
+
+Token={{ confirmation_token }}{% endblocktrans %}
+
+{% blocktrans with site_url=site_url %}If link above does not work, you can copy-paste it directly into your browser.{% endblocktrans %}
```
This line is now a bit obsolete?

mitar commented:
OK, for this comments on wording, I am commenting for you to learn what information is useful to convey when you are talking to the user. If you do not want, I can accept pull request (when you fix the rest) without fixing this and I will do this myself latter. But I think it is useful to learn. But it is your decision. I don't want you to feel frustrated about this.

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,67 @@ def dispatch(self, request, *args, **kwargs):
+    success_url = urlresolvers.reverse_lazy('account')
+
+    def form_valid(self, form):
+        subject = loader.get_template('user/confirmation_email_subject.txt')
```
Just use:

https://docs.djangoproject.com/en/dev/ref/templates/api/#the-render-to-string-shortcut

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,67 @@ def dispatch(self, request, *args, **kwargs):
+            'site_name' : settings.EMAIL_SUBJECT_PREFIX,
+        }))
+
+        from_email = settings.DEFAULT_FROM_EMAIL
```
You do not need this. This is used automatically if you do not set anything as from e-mail.

premik91 commented:
Will not be frustrated. Will fix :)

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,67 @@ def dispatch(self, request, *args, **kwargs):
+
+        # Message content
+        # We generate a random string with length 77
+        confirmation_token = ''.join(random.choice(string.letters + string.digits) for i in xrange(77))
```
Use `django.crypto.get_random_string`.

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,67 @@ def dispatch(self, request, *args, **kwargs):
+        # Message content
+        # We generate a random string with length 77
+        confirmation_token = ''.join(random.choice(string.letters + string.digits) for i in xrange(77))
+        text_content = loader.get_template('user/confirmation_email.txt')
```
Same, use `render_to_string`. ;-)

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,67 @@ def dispatch(self, request, *args, **kwargs):
+        # We generate a random string with length 77
+        confirmation_token = ''.join(random.choice(string.letters + string.digits) for i in xrange(77))
+        text_content = loader.get_template('user/confirmation_email.txt')
+        text_content = text_content.render(template.Context({
```
Use the context as I described above.

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,67 @@ def dispatch(self, request, *args, **kwargs):
+            'domain' : self.request.build_absolute_uri(urlresolvers.reverse('home'))
+        }))
+
+        user.email_confirmation_token = models.EmailConfirmationToken(value=confirmation_token, created_time=timezone.now())
```
You do not have to set created_time, it is done automatically.

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,67 @@ def dispatch(self, request, *args, **kwargs):
+
+        user.email_confirmation_token = models.EmailConfirmationToken(value=confirmation_token, created_time=timezone.now())
+        user.save()
+        user.email_user(subject, text_content, from_email)
```
No from_email.

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,67 @@ def dispatch(self, request, *args, **kwargs):
+        user.save()
+        user.email_user(subject, text_content, from_email)
+
+        messages.success(self.request, _("Email confirmation link has been sent to the e-mail address you provided."), fail_silently=True)
```
Mah, I would not use `fail_silently`. Better is to fail. ;-) I would also remove this flag elsewhere we are suing it.

Again, user has not provided anything.

_Confirmation e-mail has been sent to your e-mail address._

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,67 @@ def dispatch(self, request, *args, **kwargs):
         return form_class(self.request.user, **self.get_form_kwargs())
 
+class EmailConfirmationSendToken(edit_views.FormView):
+    template_name = 'user/email_confirmation.html'
```
`user/email_confirmation_send_token.html`

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,67 @@ def dispatch(self, request, *args, **kwargs):
+        return super(EmailConfirmationSendToken, self).form_valid(form)
+
+class EmailConfirmationProcessToken(generic_views.FormView):
+    template_name = 'user/email_confirmaton_final.html'
```
`user/email_confirmation_process_token.html`

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,67 @@ def dispatch(self, request, *args, **kwargs):
+        user = self.request.user
+        user.email_confirmed = True
+        user.save()
+        messages.success(self.request, _("You have successfully confirmed your e-mail address"), fail_silently=True)
```
Dot at the end!

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,67 @@ def dispatch(self, request, *args, **kwargs):
+    def get_initial(self):
+        if self.kwargs.has_key('confirmation_token'):
+            return {
+                'confirmation_token': self.kwargs['confirmation_token'],
```
`self.kwargs.get('confirmation_token')` and then you do not need that `has_key` check?

mitar commented on piplmesh/urls.py:
```diff
@@ -60,6 +60,9 @@
     url(r'^account/password/change/$', account_views.PasswordChangeView.as_view(), name='password_change'),
+    url(r'^account/confirmation/$', account_views.EmailConfirmationSendToken.as_view(), name='email_confirmation'),
+    url(r'^account/confirmation/token/$', account_views.EmailConfirmationProcessToken.as_view(), name='email_confirmaton_no_token'),
+    url(r'^account/confirmation/token/(?P<confirmation_token>\w+)/$', account_views.EmailConfirmationProcessToken.as_view(), name='email_confirmaton_with_token'),
```
Combine those two URLs into one. URL resolving will then do better. And user will not have to decide to use `email_confirmaton_no_token` or `email_confirmaton_with_token` when creating an URL. Simply make the confirmation token part optional.

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,67 @@ def dispatch(self, request, *args, **kwargs):
+                'confirmation_token': self.kwargs['confirmation_token'],
+            }
+
+    def dispatch(self, request, *args, **kwargs):
```
This is not good. We have to think a bit more about this. What we want to prevent or allow? We have lazy users (not authenticated) and non-lazy (those who registered or logged in through third-party authentication). In one other pull request (#198) we have decided that also lazy users will be able to set their e-mail addresses. So in fact everybody can go and send its confirmation token.

But what is important is that user has e-mail address defined! And that when user gets back, the same user is still logged in!

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,67 @@ def dispatch(self, request, *args, **kwargs):
     def get_form(self, form_class):
         return form_class(self.request.user, **self.get_form_kwargs())
 
+class EmailConfirmationSendToken(edit_views.FormView):
```
This (sending token) should be possible to do only if user has e-mail address defined.

Commit:
Fixes

Commit:
Some fixes

Commit:
Fix

Commit:
Fixed

premik91 commented:
Ok, I think I fixed everything else except these url. I am unable to figure out how to make them working(what do I have to load to make it wok). Searched a lot, but getting tired of searching, please help

mitar commented:
(Please do better commit messages.)

You know this template tag:

https://docs.djangoproject.com/en/dev/ref/templates/builtins/#url

This (and the one from django-missing) should be all you need.

mitar commented on piplmesh/urls.py:
```diff
@@ -60,6 +60,8 @@
     url(r'^account/$', account_views.AccountChangeView.as_view(), name='account'),
     url(r'^account/password/change/$', account_views.PasswordChangeView.as_view(), name='password_change'),
+    url(r'^account/confirmation/$', account_views.EmailConfirmationSendToken.as_view(), name='email_confirmation'),
+    url(r'^account/confirmation/token?/?(?P<confirmation_token>\w+)/$', account_views.EmailConfirmationProcessToken.as_view(), name='email_confirmaton'),
```
Or is this what you have problems with?

mitar replied:
`r'^account/confirmation/token/(?:(?P<confirmation_token>\w+)/)?$'` will do it for you. ;-)

mitar replied:
Please describe the regex above. For me to see that you understand it.

mitar replied:
(Not in the code comment, here.)

premik91 replied:
No mine actually worked also, so this wasn't the problem(but if it was only /token/ it return confirmation_token='n', no idea why). But love you solution. If I understand correctly, this is just to check if there is anything after <code>/token/</code>, it is actually very similiar to if sentence. If there is it will add the code after the first and until the last question mark. Else it will add nothing.

mitar replied:
The point is that your regex matched also `account/confirmation/toke/`. I do not know why you have that first `?` in there. This makes `n` optional. ;-) And then you also have `/` optional. So you get `account/confirmation/toke`, and `n` is captured into `confirmation_token` and then follows `/`. :-)

You didn't made `confirmation_token` optional. So it had to be matched. So `n` went there. And the rest was left out because it was optional. So if you want that `confirmation_token` is optional, you have to make it optional. So the best is simply to go like you want it to go. Everything until `account/confirmation/token/` is mandatory. And then the rest is optional. And how is the rest constructed? From `confirmation_token` and `/`. And this is it.

Your interpretation of regex is a bit strange. You should do not talk about adding, but about matching. Imagine like regex overlaying over the string. So you overlay `account/confirmation/token/`. And then you have one `()` section which is optional. This is made with `?` at the end. Which means zero on one match. So if will match preceding part zero or one times. And preceding thing is made from `\w+` which is captures into `confirmation_token`. And `/`. The same would be to write:

```
r'^account/confirmation/token/((?P<confirmation_token>\w+)/)?$'
```

Only then the `()` would also capture everything inside. Because we are not interested in this group, but only the part without `/`. We use special `?:` at the beginning to tell that this group should not capture anything but it is there just to combine things together. (So that last `?` can work on everything inside or nothing.)

mitar commented on piplmesh/settings.py:
```diff
@@ -49,6 +49,7 @@
 URL_VALIDATOR_USER_AGENT = 'Django'
 
 SITE_ID = 1
```
This line you can remove. We are not using that.

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,25 @@
+
+{% blocktrans %}Please click the link below to confirm your e-mail address:{% endblocktrans %}
+
+{% url account as the_url %}{% fullurl the_url %}{{ confirmation_token }}
```
Not good!

premik91 replied:
I know. That's where I have the problem. It doesn't work(fullurl). Sometin missing but can't figure out what.

mitar replied:
I am not sure, how it does not work? What happens? From what I see it of course should not work. You are doing URL to account page, not to process token page. And you should pass confirmation_token as parameter to `url` tag.

premik91 replied:
Well I don't know. Do I have to load something? <code>Invalid block tag: 'fullurl'</code>
I am not saying it doesn't make urls that I want, it doesn't work at all :)

Or am I missing something vital?

mitar replied:
Have you read that page with django-missing documentation I gave to you? There is written which template tag library you have to load and how. Read the page a bit. ;-)

mitar replied:
(You are missing reading documentation. :-) )

premik91 replied:
I read it, just didn't think missing meant that it was actually missing from django. :D

premik91 commented:
Fixes  
Some fixes  
Fix  
Fixed
Hahaha, not informative enough? That's strange :D

Commit:
Removed SITE_ID, updated urls.py

mitar commented on piplmesh/frontend/static/piplmesh/css/style.css:
```diff
@@ -258,6 +267,14 @@ input.gsc-search-button {
+    margin-bottom: 100px;
+}
+
+#email_confirmation_final input[type=text]  {
```
`final`? Didn't we rename this?

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,64 @@ def dispatch(self, request, *args, **kwargs):
+        }
+
+    def dispatch(self, request, *args, **kwargs):
+        # TODO: User has e-mail address defined! And that when user gets back, the same user is still logged in!
```
Split into two TODOs. One which tells that we have check if user has e-mail defined (verb _check_ is missing). And this should also go to the `EmailConfirmationSendToken` view.  And then the other one which should check if it is still the same user.

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,64 @@ def dispatch(self, request, *args, **kwargs):
+
+    def get_initial(self):
+        return {
+                'confirmation_token': self.kwargs.get('confirmation_token'),
```
Indentation.

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,64 @@ def dispatch(self, request, *args, **kwargs):
+        user = self.request.user
+
+        subject = loader.render_to_string('user/confirmation_email_subject.txt', {
+            'SITE_NAME' : settings.SITE_NAME,
```
Can you order this alphabetically? Constants first, variables last?

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,64 @@ def dispatch(self, request, *args, **kwargs):
+
+        # Message content
+        # We generate a random string with length 77
+        confirmation_token = crypto.get_random_string(77, string.letters + string.digits)
```
You can probably just leave to default `get_random_string` letters to chose from? Or you think we have to use `string.letters + string.digits`?

premik91 replied:
It is probably not necessary no. But it looks good :D

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,64 @@ def dispatch(self, request, *args, **kwargs):
+        # We generate a random string with length 77
+        confirmation_token = crypto.get_random_string(77, string.letters + string.digits)
+        text_content = loader.render_to_string('user/confirmation_email.txt', {
+            'SITE_NAME' : settings.SITE_NAME,
```
Order.

mitar replied:
Also add the address to which token is send.

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,64 @@ def dispatch(self, request, *args, **kwargs):
+            'CONFIRMATION_TOKEN_VALIDITY' : models.CONFIRMATION_TOKEN_VALIDITY,
+        })
+
+        user.email_confirmation_token = models.EmailConfirmationToken(value = confirmation_token)
```
No space around last `=`.

mitar commented on piplmesh/account/templates/user/email_confirmation_process_token.html:
```diff
@@ -0,0 +1,20 @@
+        <p id="current_item"><a href="{% url account %}">{% trans "Account" %}</a></p>
+    </div>
+    <p>
+        {% blocktrans with email=user.email %}You are just one step away from confirming your e-mail address. The confirmation token was sent to {{ email }}. If for some reason the input box below is empty, paste the token in manually.{% endblocktrans %}
```
The _The confirmation token was sent to {{ email }}._ part is a bit strange. If they are on this site, they obviously came through e-mail. I would leave it out and add this information to the last sentence:

_If for some reason the input box below is empty, paste the token in manually from the confirmation e-mail you have received._

mitar commented on piplmesh/account/templates/user/email_confirmation_send_token.html:
```diff
@@ -0,0 +1,29 @@
+            <button type="submit">{% trans "Click here to confirm your e-mail address" %}</button>
+        </form>
+    {% else %}
+        <p>You haven't supplied any e-mail address! Click Account to fix this.</p>
```
Don't the wording. Also if you can, supply links. Don't have users search for how to get to the solution of their problem.

BTW, I would not check this in the template. I would check in dispatch or somewhere. If e-mail is not configured, they don't have a reason to be on this page anyway.

mitar commented on piplmesh/account/templates/user/account.html:
```diff
@@ -15,6 +15,9 @@
     <div id="password_change">
-        <a href="{% url password_change %}">{% trans "Change your password" %}</a></p>
+        <p><a href="{% url password_change %}">{% trans "Change your password" %}</a></p>
+        {% if not user.email_confirmed %}
```
Here you should fix the if, if e-mail is defined and if it was not confirmed, then display the link.

Commit:
Added TODO's, minor sentence change, arranged alphabetically, renamed id CSS

premik91 commented:
Can we try to finish this today? Because I have a week full of exams and I would be able to continue this on wednesday, and I think you don't want that either :)

Commit:
Fix

Commit:
Updated translations

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,73 @@ def dispatch(self, request, *args, **kwargs):
+
+    def get_form(self, form_class):
+        return form_class(self.request.user, **self.get_form_kwargs())
+
```
Unnecessary empty line.

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,73 @@ def dispatch(self, request, *args, **kwargs):
+        })
+
+        # Message content
+        # We generate a random string with length 77
```
This two comments are not really useful. Now we have really verbose code: `crypto.get_random_string(77)`, it is quite obvious that it is a random string of length 77. ;-)

premik91 replied:
True :)

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,73 @@ def dispatch(self, request, *args, **kwargs):
+            'SITE_NAME' : settings.SITE_NAME,
+            'confirmation_token' : confirmation_token,
+            'request' : self.request,
+            'user' : user,
```
e-mail address still missing.

mitar replied:
Aha, you added it to subject, not to content. ;-)

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,73 @@ def dispatch(self, request, *args, **kwargs):
+            'CONFIRMATION_TOKEN_VALIDITY' : models.CONFIRMATION_TOKEN_VALIDITY,
+            'EMAIL_SUBJECT_PREFIX' : settings.EMAIL_SUBJECT_PREFIX,
+            'SITE_NAME' : settings.SITE_NAME,
+            'user_address' : user.email,
```
`user_address` -> `email_address`

mitar commented on piplmesh/account/views.py:
```diff
@@ -1,13 +1,14 @@
-import json, urllib, urlparse
+import json, random, string, urllib, urlparse
```
Unnecessary imports?

mitar commented on piplmesh/account/templates/user/email_confirmation_send_token.html:
```diff
@@ -0,0 +1,25 @@
+    <p>
+        {% blocktrans with email=user.email %}Clicking the button below will send a confirmation mail to {{ email }}.{% endblocktrans %}
+    </p>
+    <form method="post" action="{% url email_confirmation %}" id="email_confirmation">
```
Why are you not rendering this with `{% include "form/form.html" %}`? You can move `id="email_confirmation"` to something else.

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,73 @@ def dispatch(self, request, *args, **kwargs):
+        user = self.request.user
+
+        subject = loader.render_to_string('user/confirmation_email_subject.txt', {
+            'CONFIRMATION_TOKEN_VALIDITY' : models.CONFIRMATION_TOKEN_VALIDITY,
```
No spaces before `:`.

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,25 @@
+{% url account as the_url %}{% fullurl the_url %}{{ confirmation_token }}
+
+{% blocktrans %}Or, open the following URL:{% endblocktrans %}
+{% url account as the_url %}{% fullurl the_url %}
```
Empty lines before and after this URL.

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,25 @@
+
+{{ confirmation_token }}
+
+{% blocktrans with CONFIRMATION_TOKEN_VALIDITY=CONFIRMATION_TOKEN_VALIDITY %}The link will be active for {{ CONFIRMATION_TOKEN_VALIDITY }} days, if it expires you will have to redo the process of confirming your e-mail address.{% endblocktrans %}
```
Here you should translate based on the number `CONFIRMATION_TOKEN_VALIDITY`. If `CONFIRMATION_TOKEN_VALIDITY` is 1, then there should not be _days_, but something else. See how this is done.

Also, end a sentence before _if it expires_.

_Confirmation token will be active ..._

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,25 @@
+
+{% url account as the_url %}{% fullurl the_url %}{{ confirmation_token }}
+
+{% blocktrans %}Or, open the following URL:{% endblocktrans %}
```
_URL_ -> _link_, if you are using it elsewhere.

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,25 @@
+
+{% blocktrans with CONFIRMATION_TOKEN_VALIDITY=CONFIRMATION_TOKEN_VALIDITY %}The link will be active for {{ CONFIRMATION_TOKEN_VALIDITY }} days, if it expires you will have to redo the process of confirming your e-mail address.{% endblocktrans %}
+
+{% blocktrans %}Please also be careful to open complete URL for confirmation. Your e-mail client might have broken it into several lines.{% endblocktrans %}
```
_to open complete link._ (You do not have to repeat that we are confirming things here, this is clear.

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,25 @@
+{% blocktrans %}Yours,{% endblocktrans %}
+
+{{ SITE_NAME }}
+{% url account as the_url %}{% fullurl the_url %}
```
Account? Not home?

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,25 @@
+
+{% blocktrans %}If you have any problems confirming your e-mail address or have any other questions just reply to this e-mail.{% endblocktrans %}
+
+
```
Remove this empty line.

Commit:
Removed empty lines, restructured, including form/form.html

Commit:
Added missing url_tags.py

Commit:
Minor url fixes

Commit:
Url fixed in confirmation_email.txt template

Commit:
Added wordwrap filter

Commit:
home instead of account url

Commit:
Added plural tag

mitar commented on piplmesh/account/templatetags/url_tags.py:
```diff
@@ -0,0 +1,168 @@
```
NOOOOOOOOOOOOOOOO. You made some big copy-pasting here. You know that I don't like it. You know why libraries are? Because they can be _used_. Not because you copy-paste the code. The idea is that if library is updated, you get its update for free. Not that you have to keep things in sync.

So, django-missing is a library and you should add it to our project. How we are adding libraries to our project?

mitar replied:
I can still see in the pull request!

premik91 replied:
Add it to requirements.txt ?

premik91 replied:
From here https://bitbucket.org/mitar/django-missing/?
How do I do that from bitbucket. Do I have to clone it (hg clone ssh://hg@bitbucket.org/mitar/django-missing) or how?

mitar replied:
Yes!

premik91 replied:
So I just copy this line to requirements.txt?
<code>hg clone ssh://hg@bitbucket.org/mitar/django-missing</code>

mitar replied:
(That yes was for adding to `requirements.txt`.)

You could add Mercurial repository in a similar way how we were adding git repositories. But this library is published on PyPi. So you can just add it only with name and version and will find it automatically.

```
pip install django-missing
```

premik91 commented:
This is now done :)

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,28 @@
+{% load i18n url_tags %}{% blocktrans with username=user.username %}Hello {{ username }}!{% endblocktrans %}
+{% filter wordwrap:80 %}
```
I am not sure if you want to wrap everything at once, because what happens with long URLs? Does it split them? Or does it leave them? Because if it splits them, then you have to use this around every paragraph.

I would wordwrap at 72.

premik91 replied:
I doesn't. It leaves url's the same. Ok 72.

mitar replied:
Good.

Commit:
word-wrap moved to 72

mitar commented:
What is now done? Copy-paste? Nooo!

Commit:
Removed url_tags.py

Commit:
Added djnago-missing to requirements.txt

Commit:
Updated translations

Commit:
Merge remote-tracking branch 'upstream/master' into Issue-#196

premik91 commented:
Done?

mitar commented on requirements.txt:
```diff
@@ -13,3 +13,4 @@ django-sekizai==0.5
 django-tastypie-mongoengine==0.2.1
 -e git+https://github.com/mitar/django-mongodbforms.git@01201c7b14127a3cb2cb027965320ed8ec499a4e#egg=mongodbforms-dev
 -e git+https://github.com/mitar/django-mongogeneric.git@0d9d320398865d04e5c9d72b47543ecf68471813#egg=mongogeneric-dev
+django-missing
```
No, also version should be added. See the output of `pip freeze`.

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,28 @@
+
+{% blocktrans %}Please click the link below to confirm your e-mail address:{% endblocktrans %}
+
+{% url email_confirmaton_process_token as the_url %}{% fullurl the_url %}{{ confirmation_token }}
```
Not good. You are appending `confirmation_token` manually. See `url` tag to see how you pass this as a parameter so that it is constructed automatically.

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -6,206 +6,340 @@ msgid ""
+msgid ""
+"Please enter the confirmation token you received to your e-mail address."
+msgstr ""
+"Prosimo vnesi potrditveno kodo, ki si jo prejel na svoj e-mail naslov. "
```
Unnecessary space after `.`.

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -6,206 +6,340 @@ msgid ""
-#: .\frontend\templates\user\user.html.py:11
+#: account/views.py:353
+msgid "Confirmation e-mail has been sent to your e-mail address."
+msgstr "Potrditvena povezava je bila poslana na tvoj e-mail naslov."
```
_povezava_? E-mail?

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -6,206 +6,340 @@ msgid ""
+
+#: account/views.py:371
+msgid "You have successfully confirmed your e-mail address."
+msgstr "Uspešno si potrdil svoj e-mail naslov"
```
Dot at the end.

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -6,206 +6,340 @@ msgid ""
-#: .\frontend\templates\header.html.py:45
+#: account/templates/user/account.html:20
+msgid "Please confirm your e-mail address"
+msgstr "Prosimo, da potrdiš svoj e-mail naslov"
```
_da potrdiš_ -> _potrdi_.

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -6,206 +6,340 @@ msgid ""
+"your account on %(SITE_NAME)s with username %(username)s. If you have "
+"already done so or don't want to, you can safely ignore this e-mail."
+msgstr ""
+"To sporočilo ti je bilo poslano, da potrdiš ta e-mail naslov za "
```
_ta e-mail naslov za_ -> _ta e-mail naslov kot svojega za_

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -6,206 +6,340 @@ msgid ""
+msgstr ""
+"To sporočilo ti je bilo poslano, da potrdiš ta e-mail naslov za "
+"%(SITE_NAME)s račun z uporabniškim imenom %(username)s. Če si to že storil "
+"ali pa tega nočeš, lahko varno ignoriraš ta e-mail."
```
_varno_ -> _brez skrbi_

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -6,206 +6,340 @@ msgid ""
+
+#: account/templates/user/confirmation_email.txt:5
+msgid "Please click the link below to confirm your e-mail address:"
+msgstr "Pritisni povezavo spodaj za potrditev tvojega e-mail naslova"
```
`:` at the end. _Pritisni_ -> _Klikni spodnjo povezavo_

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -6,206 +6,340 @@ msgid ""
+"questions just reply to this e-mail."
+msgstr ""
+"Če imaš karšnekoli težave pri potrjanju svojega e-mail naslova ali kakšna "
+"druga vprašanja, le odgovori na ta e-mail."
```
_le odgovori na ta e-mail_ -> _lahko odgovoriš kar na ta e-mail_

Commit:
Translation fixes

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -6,206 +6,340 @@ msgid ""
+
+#: account/templates/user/email_confirmation_send_token.html:16
+msgid "Click here to confirm your e-mail address"
+msgstr "Pritisni tukaj za potrditev tvojega e-mail naslova"
```
_Klikni_

mitar replied:
_svojega_!

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -6,206 +6,340 @@ msgid ""
+"you've confirmed your e-mail address, you will be able to reset your "
+"password in the case you lost it or forget it."
+msgstr ""
+"Potrditev tvojega e-mail naslova poveča varnost tvojega računa.Ko boš "
```
_svojega e-mail naslova_

Space after dot.

Commit:
Translation fixes

mitar commented:
Almost done!

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -6,206 +6,340 @@ msgid ""
-#: .\frontend\templates\user\user.html.py:11
+#: account/views.py:353
+msgid "Confirmation e-mail has been sent to your e-mail address."
+msgstr "Potrditven e-mail je bil poslan na tvoj e-mail naslov."
```
Potrditven**i**

Commit:
Fixed url

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,28 @@
+{{ confirmation_token }}
+
+{% blocktrans with CONFIRMATION_TOKEN_VALIDITY=CONFIRMATION_TOKEN_VALIDITY count days=CONFIRMATION_TOKEN_VALIDITY %}Confirmation token will be active for {{ CONFIRMATION_TOKEN_VALIDITY }} day.
+{% plural %}Confirmation token will be active for {{ CONFIRMATION_TOKEN_VALIDITY }} days.{% endblocktrans %}
```
Here at the end is one unnecessary line break.

mitar replied:
I am not sure if it is OK to have a newline before `{% plural %}`.

premik91 replied:
It is so in Djnago documentation:
<code>
{% blocktrans count list|length as counter %}
There is only one {{ name }} object.
{% plural %}
There are {{ counter }} {{ name }} objects.
{% endblocktrans %}</code>

mitar replied:
Yes. But there is newline always, in both cases. You have it only in one case.

premik91 replied:
Noticed now and fixed :)

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,28 @@
+{% load i18n url_tags %}{% blocktrans with username=user.username %}Hello {{ username }}!{% endblocktrans %}
+{% filter wordwrap:72 %}
+{% blocktrans with SITE_NAME=SITE_NAME username=user.username %}This message was sent to you to confirm this e-mail address as yours for your account on {{ SITE_NAME }} with username {{ username }}. If you have already done so or don't want to, you can safely ignore this e-mail.{% endblocktrans %}
```
Put `"` around username output.

Commit:
Added version to django-missing

mitar commented on piplmesh/account/views.py:
```diff
@@ -316,6 +319,71 @@ def dispatch(self, request, *args, **kwargs):
+            'user': user,
+        })
+
+        confirmation_token = crypto.get_random_string(77)
```
OK, 77 is reaaaaly tooo much. Do 20, more than enough.

premik91 replied:
Hahaha, yeah I knew that :)

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,28 @@
+
+{% blocktrans %}If you have any problems confirming your e-mail address or have any other questions just reply to this e-mail.{% endblocktrans %}
+{% endfilter %}
+{% blocktrans %}Yours,{% endblocktrans %}
```
One empty line more here, please.

mitar replied:
Before. ;-) Not after. Sorry.

mitar replied:
Oh, ignore. I cannot decide. ;-)

premik91 replied:
Little help: ;)

Hello test!

This message was sent to you to confirm this e-mail address as yours for
your account on PiplMesh with username "test". If you have already done
so or don't want to, you can safely ignore this e-mail.

Please click the link below to confirm your e-mail address:

http://127.0.0.1:8000/account/confirmation/token/Ei3GCO3ce9rJQXUe8p7U/

Or, open the following link:

http://127.0.0.1:8000/account/confirmation/token/

and copy-paste the following confirmation token into the input box
there:

Ei3GCO3ce9rJQXUe8p7U

Confirmation token will be active for 5 days. If it expires you will
have to redo the process of confirming your e-mail address.

Please also be careful to open complete link. Your e-mail client might
have broken it into several lines.

If you have any problems confirming your e-mail address or have any
other questions just reply to this e-mail.

Yours,

PiplMesh
http://127.0.0.1:8000/

Commit:
From 77 to 20 chars

mitar commented:
I got an error using your code. Template tags library was not found. You are missing our crucial thing for them to be loaded. Please read documentation. ;-) On Django template tags libraries.

Have you tested your code lately? ;-)

mitar commented:
And, now, there is a question. We would need tests for this code. ;-) Do you want to write them. Or would you rather open a new ticket for that for your peers?

mitar commented on piplmesh/account/forms.py:
```diff
@@ -132,3 +132,35 @@ class PasswordChangeForm(UserCurrentPasswordForm, UserPasswordForm):
+
+    confirmation_token = forms.CharField(
+        label=_("Confirmation token"),
+        min_length=77,
```
Hmm. ;-)

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -6,206 +6,340 @@ msgid ""
+msgid ""
+"Please also be careful to open complete link. Your e-mail client might have "
+"broken it into several lines."
+msgstr ""
```
Missing?

premik91 replied:
<code>Prosimo bodi previden/a, da odpreš celotno povezavo. Tvoj e-poštni odjemalec jo je morda razdelil v več vrstic.</code>
Ok?

Commit:
Fixed confirmation_token form, trasnlation fixes, email message fix

Commit:
Translation fix

premik91 commented:
I will open a new ticket, so we can finish this, because I have two exams till wednesday. But if nobody will do them till then, I will gladly, would like to learn that too.

mitar commented:
OK.

mitar commented on piplmesh/account/models.py:
```diff
@@ -31,6 +32,18 @@ class Connection(mongoengine.EmbeddedDocument):
     channel_id = mongoengine.StringField()
 
+class EmailConfirmationToken(mongoengine.EmbeddedDocument):
+    value = mongoengine.StringField(max_length=77, required=True)
```
Ehm. Can you please grep for 77? ;-)

premik91 replied:
Yes sorry, was just doing that. But this one is the last

Commit:
Fixed models token leght to 20

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -6,206 +6,340 @@ msgid ""
+msgid ""
+"Please also be careful to open complete link. Your e-mail client might have "
+"broken it into several lines."
+msgstr "Prosimo bodi previden/a, da odpreš celotno povezavo. Tvoj e-poštni odjemalec jo je morda razdelil v več vrstic."
```
Mnja, we always use _e-mail_ also in Slovenian translations. Don't now start with _e-pošta_. ;-) Or do it everywhere. ;-)

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -6,206 +6,340 @@ msgid ""
+#: account/templates/user/confirmation_email.txt:3
+msgid ""
+"This message was sent to you to confirm this e-mail address as yours for "
+"your account on %(SITE_NAME)s with username %(username)s. If you have "
```
You have to rerun messages collecting every time you change a translatable string in templates. For example, here are now `"` missing around usernames.

premik91 replied:
Yes I always do, just waiting if any other fix is necessary.

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -6,206 +6,340 @@ msgid ""
-#: .\frontend\templates\header.html.py:45
+#: account/templates/user/account.html:20
+msgid "Please confirm your e-mail address"
+msgstr "Prosimo potrdi svoj e-mail naslov"
```
Comma after _prosimo_.

Commit:
Tranlsation fix

Commit:
Added comma

mitar commented:
OK. I have one task for you. Please delete all `*.pyc` files in your repository and try again to generate confirmation e-mail. ;-)

mitar commented on piplmesh/account/templates/user/confirmation_email.txt:
```diff
@@ -0,0 +1,28 @@
+{% blocktrans %}If you have any problems confirming your e-mail address or have any other questions just reply to this e-mail.{% endblocktrans %}
+{% endfilter %}
+
+{% blocktrans %}Yours,{% endblocktrans %}
```
Do it like it is done here:

https://github.com/wlanslovenija/nodewatcher/blob/master/nodewatcher/frontend/templates/registration/activation_email.txt

So move one line up, two empty lines bellow.

mitar commented on piplmesh/locale/sl/LC_MESSAGES/django.po:
```diff
@@ -6,206 +6,340 @@ msgid ""
+msgid ""
+"Confirmation token will be active for %(CONFIRMATION_TOKEN_VALIDITY)s day.\n"
+msgid_plural ""
+"Confirmation token will be active for %(CONFIRMATION_TOKEN_VALIDITY)s days.\n"
```
Because this is then messed up.

Commit:
New line to email message

Commit:
Translation fixes

mitar commented:
OK. So delete pyc files and try it. ;-)

premik91 commented:
Hmmm I see :D. Should I add this to installed apps or something else wrong?

mitar commented:
I have to go now. So there is one line change you are still missing. ;-)

mitar commented:
Yes, to installed apps. ;-)

Commit:
Added missing to apps

premik91 commented:
It doesn't work though :(. I searched around and there might be something wrong with the path, but wasn't able to repair it.

mitar commented:
Hm, but have you installed django-missing in your virtualenv? So not just adding it to requirements, but also installing it from it?

premik91 commented:
Yes I did, but there might be something wrong on my computer. So if you think this should work, it might.

I will try to remove env and try again. Nope still:

python manage.py runserver 
Validating models...

Unhandled exception in thread started by <bound method Command.inner_run of <django.contrib.staticfiles.management.commands.runserver.Command object at 0x10455a1d0>>
Traceback (most recent call last):

Commit:
Merge remote-tracking branch 'upstream/master' into Issue-#196

mitar commented:
This seems to be problem with your local installation. Google for this error and you will get some solutions.

mitar commented:
http://selenic.com/pipermail/mercurial/2007-October/015297.html

premik91 commented:
Yes thanks! It now works :)
